docs: add docs for dynamic font scaling (#3111)
ionic-team/ionic-docs: docs/layout/dynamic-font-scaling.md

In the following example we are using the [min()](https://developer.mozilla.org/en-US/docs/Web/CSS/min) function to indicate that the font size of `.foo` should be no larger than `14px`.

```css
.foo {
  font-size: min(1rem, 14px);
}
```

If the root element's default font size is `16px`, and the system-level font scale is 1.5 (i.e text sizes should be increased by 50%), then `1rem` will evaluate to `24px` because `16 * 1.5 = 24`.

This is larger than our defined maximum of `14px`, so one might assume that the evaluated font size of `.foo` is `14px`. However, since the Android Web View scales any font sizes defined using the `px` unit, this means the `14px` used in our `min()` function will also be scaled by 1.5.

As a result, this means that the maximum computed font size is actually `21px` since `14 * 1.5 = 21` and therefore the overall computed font size of `.foo` is `21px`.

### Using Modes on Different Platforms

> Each platform has slightly different font scaling behaviors, and the `'ios'` and `'md'` modes have been implemented to take advantage of the scaling behaviors on their respective platforms.

Reviewer: ```suggestion
Each platform has slightly different font scaling behaviors, and the `ios` and `md` modes have been implemented to take advantage of the scaling behaviors on their respective platforms.
```

> For example, `'ios'` mode makes use of maximum and minimum font sizes to follow [Apple's Human Interface Guidelines for Typography](https://developer.apple.com/design/human-interface-guidelines/typography). `'md'` mode does not implement this same behavior because Material Design does not have that same guidance. This means that using `'md'` mode on an iOS device may allow for very large font sizes in headers and footers.

Reviewer: ```suggestion
For example, `ios` mode makes use of maximum and minimum font sizes to follow [Apple's Human Interface Guidelines for Typography](https://developer.apple.com/design/human-interface-guidelines/typography). `md` mode does not implement this same behavior because Material Design does not have that same guidance. This means that using `md` mode on an iOS device may allow for very large font sizes in headers and footers.
```

> As a result, we strongly recommend using `'ios'` mode on iOS devices and `'md'` mode on Android devices when using Dynamic Font Scaling.

Reviewer: ```suggestion
As a result, we strongly recommend using `ios` mode on iOS devices and `md` mode on Android devices when using Dynamic Font Scaling.
```

# Dynamic Font Scaling

Dynamic Font Scaling is a feature that allows users to choose the size of the text displayed on the screen. This helps users who need larger text for better readability, and it also accommodates users who can read smaller text.

Dynamic Font Scaling is supported on Android, iOS, and iPadOS starting in Ionic v7.5.

## Try It Out

:::tip
Be sure to try this on an Android, iOS, or iPadOS device.
:::

Follow the [Changing the Font Size on a Device](#changing-the-font-size-on-a-device) guide to set your preferred font size, and watch the text in the demo below grow or shrink according to your preferences.

import DynamicFontScaling from '@site/static/usage/v7/layout/dynamic-font-scaling/index.md';

> <DynamicFontScaling />

Reviewer 2: This doesn't work for me, unfortunately 🤔 I wonder if the iframe is causing problems, since the font size adjustment doesn't make it all the way through. I'm using Chrome on a Galaxy S10e.

|Before font scaling|After font scaling |
|--------------|---------------|
|![before](https://github.com/ionic-team/ionic-docs/assets/90629384/63acad1d-3d1f-4576-afbf-e45dfab20b28)|![after](https://github.com/ionic-team/ionic-docs/assets/90629384/5aaa363f-ad1c-4e48-8056-2237774282fa)|

Author: I updated the demo so it opens in a new tab/avoids iframes: https://github.com/ionic-team/ionic-docs/pull/3111/commits/d16510a95eb09db4c2ad4e0feb2c5fa25bfaf7e3

I also added a troubleshooting step because of this in case someone tries to load Ionic in an iframe: https://github.com/ionic-team/ionic-docs/pull/3111/commits/b37af9d598dbf1072e53aee75aa00662e4e34461

Reviewer 2: We talked through this on Slack; turns out that there's a Chrome flag that's disabled by default on some devices (including mine) but is needed to make dynamic font scaling work. I confirmed that with the flag on, the playground now scales appropriately 👍 

Could we remove the troubleshooting item about not loading Ionic in an iframe, and replace it with one about ensuring the Chrome flag is enabled?

Author: Ok we demystified this a bit more, so I documented our findings here: https://github.com/ionic-team/ionic-docs/pull/3111/commits/85f54d04f17550bd917e6b63cdc6b823493f2f81

Can you give this a try again with the Accessibility Page Zoom feature enabled?

## Enabling Dynamic Font Scaling in Ionic

:::info
This feature is currently opt-in on iOS. However, it will be enabled by default starting in Ionic 8 at which point the following CSS will no longer be necessary.
:::

Dynamic Font Scaling is already enabled by default on Android. Developers can enable it on iOS by setting the following CSS globally:

```css
html {
  --ion-dynamic-font: var(--ion-default-dynamic-font);
}
```

> Under the hood, Ionic sets the following CSS on iOS devices to enable Dynamic Font Scaling:

Author: I had it there to give more insight into how this feature works, but I can remove it if you think it will be confusing for devs.

> Under the hood, Ionic sets the following CSS on iOS devices to enable Dynamic Font Scaling:

Reviewer: Maybe we could put it inside a "Note" or "Info" block so it is more of an informative note.

> Under the hood, Ionic sets the following CSS on iOS devices to enable Dynamic Font Scaling:

Author: Updated in https://github.com/ionic-team/ionic-docs/pull/3111/commits/03febb1e212a48c915e603fb12ab721cd6df289f

```css
html {
  font: var(--ion-dynamic-font);
}
```

> Developers should ensure that the [typography.css](./global-stylesheets#typographycss) file is imported.

Author: Done in https://github.com/ionic-team/ionic-docs/pull/3111/commits/7eac2a32692f22cea02a70209b19e40b9af00a23.

## Using Dynamic Font Scaling

### Integrating Custom Components

Developers can configure their custom components to take advantage of Dynamic Font Scaling by converting any pixel font sizes to use [rem units](https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Values_and_units#lengths). An easy way to convert from `px` to `rem` is to divide the pixel font size by the default browser font size, which is typically 16px.

For example, if a component has a font size of `14px`, then this could be converted to `rem` by doing `14px / 16px = 0.875rem`.

> Developers should also audit their applications and make any other changes necessary to support larger font sizes. One thing to keep in mind is that the dimensions of your components may need to change to accommodate the larger font sizes. As a result, it may be necessary to change properties such as `width` and `height` to `min-width` and `min-height`. We also recommend having long text wrap to the next line instead of truncating to keep large text readable.

Reviewer: Should we also mention `line-height` here? That seems to be one that gets set to `px` values sometimes.

Author: I mentioned `width` and `height` as an example -  I wasn't focused on calling out the most common properties. I can add it if you think it would help though.

Reviewer: What if we linked to the documentation on [length](https://developer.mozilla.org/en-US/docs/Web/CSS/length) and just specified they should look over any CSS properties that accept length and change them from `px` to `rem` units?

Author: How does https://github.com/ionic-team/ionic-docs/pull/3111/commits/a9e12380a2df5fc6ec70a70f5d3ee0cefc99cae8 sound?

Reviewer 2: This paragraph was originally less about updating px values to rem and more dealing with edge cases like the element needing to be allowed to grow in size. At the moment, the info is pretty much duplicated between this paragraph and the two above it.

Author: So my original goal with this section is to note that following:

1. Devs need to convert font size px units to rem
3. Devs may need to change other CSS values to account for the new font sizes

The change I made is meant to highlight point 3 whereas the first paragraph is meant to highlight point 1. Not every non-font-size CSS property that uses px will necessarily need to be converted to rem.

Reviewer 2: I suppose I felt like specifically mentioning that `width` and `height` may need updating to `min-width` and `min-height` was more useful, since that's not a px/rem change and seems common. It's okay how it is now though.

Reviewer: I like the addition of the link to the length docs. I think mentioning the `width`/`height` needing to change to `min-width`/`min-height` might be good for certain cases, like if they have a button with the following and want to support text wrapping:

```diff
button {
-  height: 48px;
+  min-height: 48px;
}
```

But in other cases they might have a rounded button like this that needs to scale proportionately:

```diff
button {
-  width: 48px;
-  height: 48px;
+  width: 3rem;
+  height: 3rem;
}
```

Author: OK I cleaned this up a bit. Thoughts on https://github.com/ionic-team/ionic-docs/pull/3111/commits/b0360dfb76ecd8bf971023f6b1d81d045fd61867?

Reviewer 2: Much better 👍

Author: Thanks for the help both of you!

Also note that any Ionic components that have had their font sizes overridden should also be updated to use `rem` units.

### Custom Font Family

We recommend using the default fonts in Ionic as they are designed to look good at any size and ensure consistency with other mobile apps. However, developers can use a custom font family with Dynamic Font Scaling via CSS:

```css
html {
  --ion-dynamic-font: var(--ion-default-dynamic-font);
  --ion-font-family: 'Comic Sans';
}
```

### `em` units versus `rem` units

> Developers have two options for relative font sizes: `em` and `rem`.

Reviewer: Maybe link to this section? https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Values_and_units#ems_and_rems

Author: Added in https://github.com/ionic-team/ionic-docs/pull/3111/commits/d94841fd5e6c2cb03b8c7819451db5536a168dd8

`em` units set the font size of an element relative to the font size of its parent.

> In the following example, the computed font size of `.child` is `40px` because it is a child of `.parent` (2em \* 20px = 40px).

Reviewer: ```suggestion
In the following example, the computed font size of `.child` is `40px` because it is a child of `.parent` (`2em \* 20px = 40px`).
```

```css
.parent {
  font-size: 20px;
}

.child {
  font-size: 2em;
}
```

However, the `em` unit has a compounding effect which can cause issues. In the following example, the second `.child` element has a computed font size of `80px` since the font sizes compound.

```html
<div class="parent">
  Parent element with 20px
  <div class="child">
    Child element with 40px
    <div class="child">Child element with 80px</div>
  </div>
</div>
```

> <div style={{ fontSize: '20px' }}>

Reviewer: Should we put this in a playground demo with just a javascript example? I know you don't need ionic for this, so we could override the index file and keep it simple.

Author: I think that may be overkill since that's going to show you Angular, React, and Vue code snippets which really aren't need. The purpose of this is to demonstrate how part of CSS works.

  Parent element with 20px
  <div style={{ fontSize: '2em' }}>
    Child element with 40px
    <div style={{ fontSize: '2em' }}>Child element with 80px</div>
  </div>
</div>

Due to this compounding effect, we strongly recommend using `rem` units instead of `em` units when working with Dynamic Font Scaling. `rem` units set the font size of an element relative to the font size of the root element, which is typically `<html>`. The default font size of the root element is typically `16px`.

In the following example, the computed font size of `.child` is `32px` because the font size is being computed relative to `html`, not `.parent`.

```css
.parent {
  font-size: 20px;
}

.child {
  font-size: 2rem;
}
```

## How Dynamic Font Scaling works in Ionic

> All Ionic components that define font sizes use [rem units](https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Values_and_units#lengths). This sizes the text in each component relative to the font size of the root element, which is usually the `html` element. This means that as the root element's font size changes, the text in all Ionic components scale in a consistent manner. This avoids the need to manually override each component's font size.

Reviewer: > All Ionic components that define font sizes use rem units

This is inaccurate, no? We have some components that don't scale (tabs) and others that use `em` units (icons).

Author: That's true. Clarified in https://github.com/ionic-team/ionic-docs/pull/3111/commits/338b4730b0c039518c63e0d303af9ff17c4444b2

### iOS

Dynamic Font Scaling in Ionic builds on top of an iOS feature called [Dynamic Type](https://developer.apple.com/documentation/uikit/uifont/scaling_fonts_automatically#overview). To do this, Ionic sets the [font](https://developer.mozilla.org/en-US/docs/Web/CSS/font) of the root element to an Apple-defined text style. For consistency, Ionic uses the [body](https://developer.apple.com/documentation/uikit/uifont/textstyle/1616682-body) text style.

> Using the Apple-defined text style enables Dynamic Type, allowing all text in Ionic components to scale according to the system-level preference. Note that these Apple-defined fonts only work on Apple devices. As a result, these fonts will not work on Android devices even if your app is using `'ios'` mode.

Reviewer: ```suggestion
Using the Apple-defined text style enables Dynamic Type, allowing all text in Ionic components to scale according to the system-level preference. Note that these Apple-defined fonts only work on Apple devices. As a result, these fonts will not work on Android devices even if your app is using `ios` mode.
```

We don't usually put this in quotes

> Ionic follows [Apple's Human Interface Guidelines for Typography](https://developer.apple.com/design/human-interface-guidelines/typography) when an app is in `'ios'` mode. As a result, important content is prioritized when the text size changes. This means a few things:

Reviewer: ```suggestion
Ionic follows [Apple's Human Interface Guidelines for Typography](https://developer.apple.com/design/human-interface-guidelines/typography) when an app is in `ios` mode. As a result, important content is prioritized when the text size changes. This means a few things:
```

> 1. Content in an `ion-header` or an `ion-footer` will have maximum font sizes to prioritize text in `ion-content` which is deemed more important than text in `ion-header` and `ion-footer`.

Reviewer: ```suggestion
1. Content in an `ion-header` or an `ion-footer` will have maximum font sizes to prioritize content in `ion-content` which is deemed more important than content in the `ion-header` and `ion-footer`.
```

Content other than text can be in all of these elements.

2. Components such as `ion-badge` and `ion-back-button` will have minimum font sizes so they remain readable.
3. Text in components such as `ion-tab-bar` and `ion-picker` do not participate in Dynamic Font Scaling according to Apple's Human Interface Guidelines.

> ### Android

Reviewer: I'm confused by this entire section. We aren't using `min`, `max` or `clamp` on Android values in Ionic and this section is under `How Dynamic Font Scaling works in Ionic`. Would this be better suited as info under `Using Dynamic Font Scaling` or a section with info on how the different devices / operating systems handle dynamic scaling?

Author: Our APIs are built on top of the browser behavior, so how Ionic behaves is influenced by the browser behavior which is why I had this section here. I mentioned min/max/clamp because developers potentially will need to use additional CSS to ensure their custom components support Dynamic Font Scaling. I wanted to call it out because developers may think the min/max/clamp behavior is an Ionic bug.

The Android Web View's font scaling mechanism is always enabled in web content and will automatically scale font sizes defined using the `px` unit. This means that any maximum or minimum font sizes specified using `px` will still be scaled even if the final font size does not align with the maximum or minimum font sizes specified.

In the following example we are using the [min()](https://developer.mozilla.org/en-US/docs/Web/CSS/min) function to indicate that the font size of `.foo` should be no larger than `14px`.

```css
.foo {
  font-size: min(1rem, 14px);
}
```

> If the root element's default font size is `16px`, and the system-level font scale is 1.5 (i.e text sizes should be increased by 50%), then `1rem` will evaluate to `24px` because `16 * 1.5 = 24`.

Reviewer: ```suggestion
If the root element's default font size is `16px`, and the system-level font scale is `1.5` (i.e text sizes should be increased by 50%), then `1rem` will evaluate to `24px` because `16 * 1.5 = 24`.
```

For example, `'ios'` mode makes use of maximum and minimum font sizes to follow [Apple's Human Interface Guidelines for Typography](https://developer.apple.com/design/human-interface-guidelines/typography). `'md'` mode does not implement this same behavior because Material Design does not have that same guidance. This means that using `'md'` mode on an iOS device may allow for very large font sizes in headers and footers.

As a result, we strongly recommend using `'ios'` mode on iOS devices and `'md'` mode on Android devices when using Dynamic Font Scaling.

## Changing the Font Size on a Device

Font scaling preferences are configured on a per-device basis by the user. This allows the user to scale the font for all applications that support this behavior. This guide shows how to enable font scaling for each platform.

### iOS

Font scaling on iOS can be configured in Settings > Accessibility > Display & Text Size > Larger Text. The slider on that page allows users to make the text larger or smaller across all apps that support font scaling.

Users can also tap the "Large Accessibility Sizes" switch to access more font scale options.

### Android

> Font scaling on Android can be configured in Settings > Accessibility > Display size and text. The "Font size" slider on that page allows users to make the text larger or smaller across all apps that support font scaling.

Reviewer 2: This varies widely across devices; for example, on my phone it's Settings > Accessibility > Visibility enhancements > Font size and style. I also worry that we'll need to maintain these steps as OS updates come out. Maybe we should just say it can be updated in the accessibility settings and leave it at that?

Author: Yep that's fair. Fixed in https://github.com/ionic-team/ionic-docs/pull/3111/commits/80e48a6aa0c1d7731e399af1b2d993343ca068d1

Reviewer 2: Should we also update the iOS description, so we don't need to worry about maintaining that either? I don't know how stable that sort of thing usually is on iPhones 😆

Author: Good idea: https://github.com/ionic-team/ionic-docs/pull/3111/commits/c105df8a27fceb2ec8b0d3c65754010761018f50

## Troubleshooting

### Dynamic Font Scaling is not working

> There are a number of reasons why Dynamic Font Scaling may not have any effect on an app . The following list, while not exhaustive, provides some things to check to debug why Dynamic Font Scaling is not working.

Reviewer 2: ```suggestion
There are a number of reasons why Dynamic Font Scaling may not have any effect on an app. The following list, while not exhaustive, provides some things to check to debug why Dynamic Font Scaling is not working.
```

## Try It Out

:::tip
Be sure to try this on an Android, iOS, or iPadOS device.
:::

Follow the [Changing the Font Size on a Device](#changing-the-font-size-on-a-device) guide to set your preferred font size, and watch the text in the demo linked below grow or shrink according to your preferences.

<a href="/docs/usage/v7/layout/dynamic-font-scaling/demo.html" target="_blank">
  Launch Dynamic Font Scaling Demo
</a> (opens in a new tab)

## Enabling Dynamic Font Scaling in Ionic

:::info
This feature is currently opt-in on iOS. However, it will be enabled by default starting in Ionic 8 at which point the following CSS will no longer be necessary.
:::

Dynamic Font Scaling is already enabled by default on Android. Developers can enable it on iOS by setting the following CSS globally:

```css
html {
  --ion-dynamic-font: var(--ion-default-dynamic-font);
}
```

Under the hood, Ionic sets the following CSS on iOS devices to enable Dynamic Font Scaling:

```css
html {
  font: var(--ion-dynamic-font);
}
```

Developers should ensure that the [typography.css](./global-stylesheets#typographycss) file is imported.

## Using Dynamic Font Scaling

### Integrating Custom Components

> Developers can configure their custom components to take advantage of Dynamic Font Scaling by converting any pixel font sizes to use [rem units](https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Values_and_units#lengths). An easy way to convert from `px` to `rem` is to divide the pixel font size by the default browser font size, which is typically 16px.

Reviewer: ```suggestion
Developers can configure their custom components to take advantage of Dynamic Font Scaling by converting any pixel font sizes to use [rem units](https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Values_and_units#lengths). An easy way to convert from `px` to `rem` is to divide the pixel font size by the default browser font size, which is typically `16px`.
```

See [how font scaling works on Android](#android) for more information.

### Font sizes are larger/smaller even with Dynamic Font Scaling disabled

Ionic components define font sizes using [rem units](https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Values_and_units#lengths) even when Dynamic Font Scaling is disabled. This sizes the text in each component relative to the font size of the root element, which is usually the `html` element. As a result, if the font size of `html` changes, the computed font size of all Ionic components will change too.

### Scaled Ionic iOS component font sizes do not exactly match native iOS equivalents

Certain native iOS components such as the Action Sheet make use of private font scales that Ionic does not have access to. While we try to stay as close as possible to the native behavior, text in some components may render slightly larger or smaller than their native counterparts.

### The text size in my Ionic app on iOS changed when enabling Dynamic Font Scaling

> The root element's default font size is typically 16px. However, Dynamic Font Scaling on iOS devices make use of the ["Body" text style](https://developer.apple.com/design/human-interface-guidelines/typography#Specifications) which has a default font size of 17px. Since the text in Ionic components is scaled relative to the root element's font size, some text may get larger or smaller when Dynamic Font Scaling is enabled, even if the system-level text scale did not change.

Reviewer: ```suggestion
The root element's default font size is typically `16px`. However, Dynamic Font Scaling on iOS devices make use of the ["Body" text style](https://developer.apple.com/design/human-interface-guidelines/typography#Specifications) which has a default font size of `17px`. Since the text in Ionic components is scaled relative to the root element's font size, some text may get larger or smaller when Dynamic Font Scaling is enabled, even if the system-level text scale did not change.
```

:::info

> iOS provides a "Callout" text style which has a default font size of 16px. However, this font style is currently not exposed to web content. See [the supported text styles in WebKit](https://webkit.org/blog/3709/using-the-system-font-in-web-content/) for more information.

Reviewer: ```suggestion
iOS provides a "Callout" text style which has a default font size of `16px`. However, this font style is currently not exposed to web content. See [the supported text styles in WebKit](https://webkit.org/blog/3709/using-the-system-font-in-web-content/) for more information.
```

Developers can test this behavior by enabling the experimental "Accessibility Page Zoom" feature in `chrome://flags`.

See https://bugs.chromium.org/p/chromium/issues/detail?id=645717 for more information.

### Using Modes on Different Platforms

Each platform has slightly different font scaling behaviors, and the `ios` and `md` modes have been implemented to take advantage of the scaling behaviors on their respective platforms.

For example, `ios` mode makes use of maximum and minimum font sizes to follow [Apple's Human Interface Guidelines for Typography](https://developer.apple.com/design/human-interface-guidelines/typography). `md` mode does not implement this same behavior because Material Design does not have that same guidance. This means that using `md` mode on an iOS device may allow for very large font sizes in headers and footers.

> As a result, we strongly recommend using `ios` mode on iOS devices and `md` mode on Android devices when using Dynamic Font Scaling.

Reviewer: > when using Dynamic Font Scaling.

Is there going to be a way for them to disable it when we make this the default?

Reviewer: Maybe we can add this in once it becomes the default in Ionic 8
